github/src/main/kotlin/gropius/sync/github/Outgoing.kt

```diff
@@ -205,21 +283,108 @@ class Outgoing(
                 issueInfo,
                 finalBlock.map { it.lastModifiedBy().value })
         }
-        if (relevantTimeline.last() is ClosedEvent) {
-            collectedMutations += githubCloseIssue(
-                imsProjectConfig,
+        if (shouldSyncType<ClosedEvent, ReopenedEvent>(
+                finalBlock, relevantTimeline, false
+            )
+        ) {
+            collectedMutations += githubCloseIssue(imsProjectConfig,
                 issueInfo,
                 finalBlock.map { it.lastModifiedBy().value })
         }
         return collectedMutations
     }
 
+    /**
+     * Check if TimelineItem should be synced or ignored
+     * @param AddingItem Item type with the same semantic as the item to add
+     * @param RemovingItem Item type invalidating the AddingItem
+     * @param finalBlock the last block of similar items that should be checked for syncing
+     * @param relevantTimeline Sorted part of the timeline containing only TimelineItems interacting with finalBlock
+     * @param restoresDefaultState if the timeline item converges the state of the issue towards the state of an empty issue
+     * @return true if and only if there are unsynced changes that should be synced to GitHub
+     */
+    private suspend inline fun <reified AddingItem : TimelineItem, reified RemovingItem : TimelineItem> shouldSyncType(
+        finalBlock: List<TimelineItem>, relevantTimeline: List<TimelineItem>, restoresDefaultState: Boolean
+    ): Boolean {
+        if (finalBlock.last() is AddingItem) {
+            val lastNegativeEvent = relevantTimeline.filterIsInstance<AddingItem>()
+                .lastOrNull { timelineEventInfoRepository.findByNeo4jId(it.rawId!!)?.githubId != null }
+            if (lastNegativeEvent == null) {
+                return !restoresDefaultState
+            } else {
+                if (relevantTimeline.filterIsInstance<RemovingItem>()
+                        .filter { it.lastModifiedAt > lastNegativeEvent.lastModifiedAt }
+                        .firstOrNull { timelineEventInfoRepository.findByNeo4jId(it.rawId!!)?.githubId != null } == null
+                ) {
+                    return true
+                }
+            }
+        }
+        return false
+    }
+
+    /**
+     * Mutate the labels of an issue upto GitHub
+     * @param imsProjectConfig active config
+     * @param issueInfo info of the issue containing the timeline item
+     * @param timeline TimelineItems for this issue
+     * @return List of functions that contain the actual mutation executors
+     */
+    private suspend fun pushLabels(
+        imsProjectConfig: IMSProjectConfig, issueInfo: IssueInfo, timeline: List<TimelineItem>
+    ): List<suspend () -> Unit> {
+        val groups = timeline.filter { (it is AddedLabelEvent) || (it is RemovedLabelEvent) }.groupBy {
+            when (it) {
+                is AddedLabelEvent -> {
+                    it.addedLabel().value
+                }
+
```

Comment:
maybe remove these empty lines

Reply:
I CAN'T! FORMATTER REVERTS THIS

Reply:
b6657f04f5a2753892bf920dc80b7bd4b08681f9